### PR TITLE
Managing Merge Conflicts course

### LINKED ISSUE #576: New Course: Managing Merge Conflicts
## Overview
**TL;DR**
Build a course identifying how to manage merge conflicts **and** include examples in different IDEs. Ideally finished before the IDE Week of webcasts (September 11 - 15) so it can be plugged during the talks. 

#### Initial Outline / Ideas

- [ ] Describe merge conflicts
- [ ] Have pre-created repository with existing merge conflicts
    - [ ] Maybe build a GitHub pages resume with existing conflicts so post-class, the students can potentially have a resume template already hosted on GitHub Pages.
- [ ] Walkthrough of handling the merge conflict(s)
- [ ] Fork or teacherbot creates a new repo for every student.
- [ ] IDEs
     - [ ] Atom
     - [ ] VS Studio Extension
     - [ ] VS Code
     - [ ] Eclipse
     - [ ] GitHub Desktop

#### Reasoning
While creating new content for the GitHub for Windows course in #573 I began creating an example of managing a merge conflict using the GitHub Extension for Visual Studio. While describing the exercise to @crichID , it was suggested that an entirely new course was created with examples of managing merge conflicts in different IDEs. 

### How to Contribute
- [ ] Comment on this issue if you'd like to work/collaborate on it! 
- [ ] Fork the [repository](https://github.com/github/training-kit).
- [ ] Use the [GitHub Flow](https://guides.github.com/introduction/flow/) to make changes to your fork. [This](https://services.github.com{{site.baseurl}}/intro-to-github/) is a refresher course if you're unsure about how to make a change on GitHub.
- [ ] Push your changes to your repository. 
- [ ] Submit a Pull Request
  - Base Dropdown: github/training-kit
  - Compare Dropdown: Your fork

#### Testing Locally
If you'd like to make and test changes locally (and see how they would look if merged), do the following:
- In your command line:
  - [Install Git](https://git-scm.com/)
  - [Install Ruby](https://www.ruby-lang.org/en/documentation/installation/)
  - [ ] Run `script/setup`
  - [ ] Run `script/server`.
    - When successful, the script will initiate a local server at `http://127.0.0.1:4000{{site.baseurl}}`.

### Questions?
- Leave a comment on this issue! Make sure to use @ mentions if you want a specific person's attention!

### PR DESCRIPTION
**Pull request by beardofedu**
## Overview
**TL;DR**
This pr closes #576 

### Testing the Examples

1. For this repo for course: https://github.com/githubschool/on-demand-merge-conflict
1. Follow the example instructions.
1. Conflict is up

#### Content Creation

- [x] Create course repo with three examples
   - [x] course repo (https://github.com/githubschool/on-demand-merge-conflict)
   - [x] **Exercise 1:** Modify config.yml with Bob Belcher's information and merge it. Create a new branch called Example 1 for the participant to edit. **Result:** Shows them a basic multi-line merge conflict in 1 file
   - [x] **Exercise 2:** Modify _data/experience.yml's first and third elements and merge it. Create a new branch called Example 2 for the participant to edit. **Result:** Shows them a merge conflict with 2 merge conflicts in the same file.
   - [x] **Exercise 3:** Modify _data/interests.yml and _data/education.yml and merge it. Create a new branch called Example 3 for the participant to edit. **Result:** Shows them a merge conflict in 2 files.
- [ ] Create course content
   - [x] Course intro
     - [x] include info about this being a GH pages resume
   - [x] Describe a normal merge
     - [x] exercise to edit config.yml lines 2-3 to make it their own
   - [x] Describe a merge conflict
   - [x] Example 1 instructions
      - [x] gif created
   - [x] Example 2 instructions
      - [x] gif created
   - [x] Example 3 instructions
      - [x] gif created
- [x] IDE Description
- [x] Setup for IDE walkthroughs (essentially identify this is a walkthrough with Example 1 / x-ref on Example 1 to this topic). The following IDEs have been identified for the initial release of the course:
   - [x] Atom
      - [x] gif created
   - [x] VS Code
      - [x] gif created
   - [x] Visual Studio Extension
      - [x] gif created
   - [ ] Eclipse
      - [ ] gif created
   - [ ] GH Desktop
      - [ ] gif created
- [x] What Next / Congratulations topic.

#### Non-content Tasks
- [ ] Create survey
- [x] Add course to `timeline`
- [x] Nav list matches course flow / implemented correctly

#### Editing / Review
- [x] Nav titles / Topic titles match
- [ ] Content free of spelling / grahamer errors
- [ ] 

### Review
This thing is ready for any final 👀

### CODE REVIEW

**Comment by beardofedu** [comment]
Exercise 1: Modify `config.yml` with Bob Belcher's information and merge it. Create a new branch called `Example 1` for the participant to edit. 

Result: Shows them a basic multi-line merge conflict in 1 file

Exercise 2: Modify `_data/experience.yml`'s first and third elements and merge it. Create a new branch called `Example 2` for the participant to edit. 

Result: Shows them a merge conflict with 2 merge conflicts in the same file.

Exercise 3: Modify `_data/interests.yml` and `_data/education.yml` and merge it. Create a new branch called `Example 3` for the participant to edit.

Result: Shows them a merge conflict in 2 files.

**Comment by beardofedu** [comment]
Concerns:

I am currently getting this error for all of the files I created: `Error reading file /Users/beardofedu/GitHub/training-kit/paths/merge-conflict/13-what-next.md: (<unknown>): did not find expected key while parsing a block mapping at line 2 column 1 `

As far as I can tell, the examples aren't forcing the students to use the command line to handle the merge conflicts, so everything can be done in the GitHub UI.

**Comment by crichID** [comment]
> As far as I can tell, the examples aren't forcing the students to use the command line to handle the merge conflicts, so everything can be done in the GitHub UI.

@beardofedu yeah, the webUI does a pretty good job with most conflicts. To reliably get a merge conflict, you need to rename the file or delete it on the master branch :wink:

**Comment by brianamarie** [comment]
@beardofedu I've gone through the merge conflicts and I have one big question - is there Teacherbot involvement that will do the automated parts of this? Or, is it meant to all be within the forked repo and the user will participate with the branches? 

In the outline above, it sets it up so the repo would have merged the branches that are supposed to be in conflict. Then there wouldn't be any conflict.  If we set it up so there are already pre-baked conflicts, we wouldn't be able to merge back into `gh-pages` with the conflicts but the user would be able to walk through everything independently. (PS we could make this super cool with Teacherbot automagically making conflicts in the future and that may be what you're already thinking.) 

#### Overall Notes
- Change ownership _(I'm happy to do this, but I'm not sure if we A) should since it's open enrollment and B) where it would go)_
- Could we have some merge conflicts happen in merges where other changes exist (AKA prep for the questions "Why is there a merge conflict on this merge, but not other merges?" "Do merge conflicts happen on every merge?" "What if most of my stuff is fine but there's only one line that actually is conflicting?") 

@beardofedu How can I help next?

**Comment by beardofedu** [comment]
Idea for how the course _might_ look:

1. Student comments in an issue prompting teacherbot to create a merge-conflict-username repository giving the student collaborator access. This repository comes with Bob, Gene, and Louise (BGL)  branches with the changes they made ready to go.
2. The repository is populated with an issue that reiterates the instructions for the student to create branches named username-experience, username-interests, and username-something (TBD)
3. When each branch is created by the student, the related BGL branch is merged into master.
4. The student follows the instructions to edit files | create a PR | resolve the merge conflict
     4.1 rinse | repeat for the other BGL examples
5. Student finishes the course. Instructed to Fork the repository so they can have their own resume GH Pages site.

cc: @brianamarie

**Comment by crichID** [comment]
> Student comments in an issue prompting teacherbot to create a merge-conflict-username repository giving the student collaborator access.

@beardofedu the only hiccup here may be that most of teacherbot's permissions exist at the repository level. I'm not sure teacherbot can create new repos in an org. I think the only org level permissions granted to bots are adding org members and working with projects.

**Comment by brianamarie** [comment]
> the only hiccup here may be that most of teacherbot's permissions exist at the repository level. I'm not sure teacherbot can create new repos in an org. I think the only org level permissions granted to bots are adding org members and working with projects.

@github/services-training Of the following paths forward, which do you think would be best for this course? Are there other options that I'm not thinking of?

1. We create a script that does this, and have it run on a Heroku server like we were doing before using Teacherbot 
2. We ask the team in charge of these permissions (I'm sure @bkeepers could point us in the right direction) if they would consider broadening the scope of permissions
3. We use a forking model, but then proceed with no Teacherbot interaction, and have users merge with other branches instead of back to `gh-pages`

**Comment by JasonEtco** [comment]
As a "GitHub App", Teacherbot cannot do anything unless its installed in a Repo, so it can't make Repos either :( 

> We create a script that does this, and have it run on a Heroku server like we were doing before using Teacherbot

That's totally doable, but I don't think there's an automated way to install Teacherbot into that new Repo. They'd have to install it themselves (which could be a one-click action)

**Comment by bkeepers** [comment]
> We ask the team in charge of these permissions (I'm sure @bkeepers could point us in the right direction) if they would consider broadening the scope of permissions

@kytrinyx would be a good person to talk to about options here.

**Comment by kytrinyx** [comment]
> As a "GitHub App", Teacherbot cannot do anything unless its installed in a Repo, so it can't make Repos either

GitHub Apps that have "admin" write permission and are installed on an organization are allowed to create repositories, as far as I can tell. This is enabled for GitHub Apps: https://developer.github.com/v3/repos/#create (see the little blue info icon next to the title of that section).

> > the only hiccup here may be that most of teacherbot's permissions exist at the repository level. I'm not sure teacherbot can create new repos in an org.

Where is teacherbot installed?

> We ask the team in charge of these permissions (I'm sure @bkeepers could point us in the right direction) if they would consider broadening the scope of permissions

The answer to this one is probably "no", but I'd like to better understand what you're trying to achieve, since I suspect that's possible without broadening the scope of permissions.

**Comment by hectorsector** [comment]
If we are considering forking, this seems doable:

1. User forks
1. User clicks link on README to install Teacherbot
1. User accepts Teacherbot permissions
1. Teacherbot [creates conflict branches](https://developer.github.com/v3/git/refs/#create-a-reference) in the fork
1. Teacherbot [creates PR]s(https://developer.github.com/v3/pulls/#create-a-pull-request) for each conflict.

**Comment by beardofedu** [comment]
Add instructions for further customizations. For example:

image

Is where you edit your avatar image.

**Comment by brianamarie** [comment]
@beardofedu I have gone through and made some changes for wording, and having them merge their pull requests at the end of each conflict resolution. Fantastic course!!! I think it's ready to 🚢 once the example repository is set up in the right spot. 👍

**Comment by beardofedu** [comment]
So here it is. The final stand. The course is done. I need some 👀 on the GitHub Extension for Visual Studio as I added that after all of the other reviews (pinging @brianamarie because you asked if there was anything you could do in the morning). I would love to BIG GREEN this tomorrow. I appreciate everyone's help on this project, I think I have gotten some form of 👀 / 👐  from everyone on the team. 

- [ ] If anyone _wants_ to create an Eclipse or GitHub Desktop merge conflict gif, just run through the the `_config.yml` section (aka Bob's part). 
- [ ] I thought about adding Bob's Burgers gifs to the individual IDE / Editor sections as a celebration of doing it within an editor.  
- [ ] BIG GREEN TIME

Thanks again for all the help. 😍 

cc: @hectorsector @hollenberry @crichID for notification things.
